File: go/pkg/chatgear/server_port.go
package chatgear

import (
	"context"
	"sync"
	"time"

	"github.com/haivivi/giztoy/pkg/audio/opusrt"
	"github.com/haivivi/giztoy/pkg/audio/pcm"
)

// ServerPort implements ServerPortRx and ServerPortTx as a bidirectional audio port.
// It manages audio input buffering, output mixing, state/stats tracking, and device commands.
type ServerPort struct {
	tx      DownlinkTx
	context context.Context
	cancel  context.CancelFunc
	logger  Logger

	// Input - audio from device
	input *opusrt.RealtimeBuffer

	// Output - mixer for audio to device
	mixer      *pcm.Mixer
	background *pcm.TrackCtrl
	foreground *pcm.TrackCtrl
	overlay    *pcm.TrackCtrl

	// Stats & State
	mu        sync.RWMutex
	gearStats *GearStatsEvent
	gearState *GearStateEvent
	closed    bool // protected by mu, prevents sending to closed channels

	// Events
	statsChanges chan *GearStatsChanges
	stateEvents  chan *GearStateEvent

	// Background goroutine tracking
	wg sync.WaitGroup
}

// NewServerPort creates a new ServerPort for the given DownlinkTx.
func NewServerPort(ctx context.Context, tx DownlinkTx) *ServerPort {
	ctx, cancel := context.WithCancel(ctx)
	p := &ServerPort{
		tx:      tx,
		context: ctx,
		cancel:  cancel,
		logger:  DefaultLogger(),

		input:        opusrt.RealtimeFrom(opusrt.NewBuffer(2 * time.Minute)),
		mixer:        pcm.NewMixer(pcm.L16Mono24K),
		statsChanges: make(chan *GearStatsChanges, 32),
		stateEvents:  make(chan *GearStateEvent, 32),
	}
	p.wg.Add(1)
	go func() {
		defer p.wg.Done()
		p.streamingOutputLoop()
	}()
	return p
}

// --- Handle (Input) ---

// Frame reads the next Opus frame from the device.
// Implements opusrt.FrameReader.
func (p *ServerPort) Frame() (opusrt.Frame, time.Duration, error) {
	return p.input.Frame()
}

// HandleOpusFrames handles incoming Opus frames from the device.
func (p *ServerPort) HandleOpusFrames(stampedOpusFrame []byte) {
	if _, err := p.input.Write(stampedOpusFrame); err != nil {
		p.logger.ErrorPrintf("handle opus frames: %v", err)
	}
}

// HandleGearStatsEvent handles incoming stats events from the device.
func (p *ServerPort) HandleGearStatsEvent(gse *GearStatsEvent) {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.closed {
		return
	}

	var gsc *GearStatsChanges
	if p.gearStats != nil {
		gsc = p.gearStats.MergeWith(gse.Clone())
	} else {
		p.gearStats = gse.Clone()
	}

	if gsc == nil {
		return
	}

	select {
	case p.statsChanges <- gsc:
	default:
		p.logger.WarnPrintf("stats changes channel is full, drop stats event")
	}
}

// HandleGearStateEvent handles incoming state events from the device.
func (p *ServerPort) HandleGearStateEvent(gse *GearStateEvent) {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.closed {
		return
	}

	// Filter out-of-order events
	if p.gearState != nil && gse.Time.Before(p.gearState.Time) {
		return
	}

	p.gearState = gse.Clone()

	select {
	case p.stateEvents <- gse.Clone(): // independent clone to avoid sharing with p.gearState
	default:
		p.logger.WarnPrintf("state events channel is full, drop state event")
	}
}

// --- Track (Output) ---

// NewBackgroundTrack creates a new background audio track.
func (p *ServerPort) NewBackgroundTrack() (pcm.Track, *pcm.TrackCtrl, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	w, ctrl, err := p.mixer.CreateTrack()
	if err != nil {
		return nil, nil, err
	}
	ctrl.SetGain(0.1)
	if p.background != nil {
		p.background.CloseWithError(nil)
	}
	p.background = ctrl
	ctrl.SetFadeOutDuration(time.Second)
	return w, ctrl, nil
}

// NewForegroundTrack creates a new foreground audio track.
func (p *ServerPort) NewForegroundTrack() (pcm.Track, *pcm.TrackCtrl, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	w, ctrl, err := p.mixer.CreateTrack()
	if err != nil {
		return nil, nil, err
	}
	if p.foreground != nil {
		p.foreground.CloseWithError(nil)
	}
	p.foreground = ctrl
	ctrl.SetFadeOutDuration(200 * time.Millisecond)
	return w, ctrl, nil
}

// NewOverlayTrack creates a new overlay audio track.
func (p *ServerPort) NewOverlayTrack() (pcm.Track, *pcm.TrackCtrl, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	w, ctrl, err := p.mixer.CreateTrack()
	if err != nil {
		return nil, nil, err
	}
	if p.overlay != nil {
		p.overlay.CloseWithError(nil)
	}
	p.overlay = ctrl
	ctrl.SetFadeOutDuration(time.Second)
	return w, ctrl, nil
}

// BackgroundTrackCtrl returns the current background track controller.
func (p *ServerPort) BackgroundTrackCtrl() *pcm.TrackCtrl {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.background
}

// ForegroundTrackCtrl returns the current foreground track controller.
func (p *ServerPort) ForegroundTrackCtrl() *pcm.TrackCtrl {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.foreground
}

// OverlayTrackCtrl returns the current overlay track controller.
func (p *ServerPort) OverlayTrackCtrl() *pcm.TrackCtrl {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.overlay
}

// Interrupt stops all output tracks immediately.
func (p *ServerPort) Interrupt() {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.background != nil {
		p.background.CloseWithError(nil)
		p.background = nil
	}
	if p.foreground != nil {
		p.foreground.CloseWithError(nil)
		p.foreground = nil
	}
	if p.overlay != nil {
		p.overlay.CloseWithError(nil)
		p.overlay = nil
	}
}

// --- Stats (getter) ---

// StatsChanges returns a channel that receives stats change events.
func (p *ServerPort) StatsChanges() <-chan *GearStatsChanges {
	return p.statsChanges
}

// StateEvents returns a channel that receives state events.
func (p *ServerPort) StateEvents() <-chan *GearStateEvent {
	return p.stateEvents
}

// Context returns the port's context.
func (p *ServerPort) Context() context.Context {
	return p.context
}

// GearStats returns the current gear stats.
func (p *ServerPort) GearStats() (*GearStatsEvent, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.gearStats, p.gearStats != nil
}

// GearState returns the current gear state.
func (p *ServerPort) GearState() (*GearStateEvent, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.gearState, p.gearState != nil
}

// Volume returns the current volume percentage.
func (p *ServerPort) Volume() (int, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.Volume != nil {
		return int(p.gearStats.Volume.Percentage), true
	}
	return 0, false
}

// LightMode returns the current light mode.
func (p *ServerPort) LightMode() (string, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.LightMode != nil {
		return p.gearStats.LightMode.Mode, true
	}
	return "", false
}

// Brightness returns the current brightness percentage.
func (p *ServerPort) Brightness() (int, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.Brightness != nil {
		return int(p.gearStats.Brightness.Percentage), true
	}
	return 0, false
}

// WifiNetwork returns the current connected WiFi network.
func (p *ServerPort) WifiNetwork() (*ConnectedWifi, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.WifiNetwork != nil {
		return p.gearStats.WifiNetwork, true
	}
	return nil, false
}

// WifiStore returns the stored WiFi list.
func (p *ServerPort) WifiStore() (*StoredWifiList, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.WifiStore != nil {
		return p.gearStats.WifiStore, true
	}
	return nil, false
}

// Battery returns the current battery status.
func (p *ServerPort) Battery() (pct int, isCharging bool, ok bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.Battery != nil {
		return int(p.gearStats.Battery.Percentage), p.gearStats.Battery.IsCharging, true
	}
	return 0, false, false
}

// SystemVersion returns the current system version.
func (p *ServerPort) SystemVersion() (string, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.SystemVersion != nil {
		return p.gearStats.SystemVersion.CurrentVersion, true
	}
	return "", false
}

// Cellular returns the current cellular network.
func (p *ServerPort) Cellular() (*ConnectedCellular, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.Cellular != nil {
		return p.gearStats.Cellular, true
	}
	return nil, false
}

// PairStatus returns the current pair status.
func (p *ServerPort) PairStatus() (string, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.PairStatus != nil {
		return p.gearStats.PairStatus.PairWith, true
	}
	return "", false
}

// ReadNFCTag returns the last read NFC tags.
func (p *ServerPort) ReadNFCTag() (*ReadNFCTag, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.ReadNFCTag != nil {
		return p.gearStats.ReadNFCTag, true
	}
	return nil, false
}

// Shaking returns the current shaking level.
func (p *ServerPort) Shaking() (float64, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if p.gearStats != nil && p.gearStats.Shaking != nil {
		return p.gearStats.Shaking.Level, true
	}
	return 0, false
}

// --- Device Commands ---

// SetVolume sets the volume of the device.
func (p *ServerPort) SetVolume(volume int) error {
	cmd := SetVolume(volume)
	return p.tx.IssueCommand(p.context, &cmd, time.Now())
}

// SetLightMode sets the light mode of the device.
func (p *ServerPort) SetLightMode(mode string) error {
	cmd := SetLightMode(mode)
	return p.tx.IssueCommand(p.context, &cmd, time.Now())
}

// SetBrightness sets the brightness of the device.
func (p *ServerPort) SetBrightness(brightness int) error {
	cmd := SetBrightness(brightness)
	return p.tx.IssueCommand(p.context, &cmd, time.Now())
}

// SetWifi sets the WiFi network of the device.
func (p *ServerPort) SetWifi(ssid, password string) error {
	return p.tx.IssueCommand(p.context, &SetWifi{SSID: ssid, Password: password}, time.Now())
}

// DeleteWifi deletes a stored WiFi network.
func (p *ServerPort) DeleteWifi(ssid string) error {
	return p.tx.IssueCommand(p.context, DeleteWifi(ssid), time.Now())
}

// Reset resets the device.
func (p *ServerPort) Reset() error {
	return p.tx.IssueCommand(p.context, &Reset{}, time.Now())
}

// Unpair unpairs the device.
func (p *ServerPort) Unpair() error {
	return p.tx.IssueCommand(p.context, &Reset{Unpair: true}, time.Now())
}

// Sleep puts the device to sleep.
func (p *ServerPort) Sleep() error {
	return p.tx.IssueCommand(p.context, &Halt{Sleep: true}, time.Now())
}

// Shutdown shuts down the device.
func (p *ServerPort) Shutdown() error {
	return p.tx.IssueCommand(p.context, &Halt{Shutdown: true}, time.Now())
}

// RaiseCall raises a call on the device.
func (p *ServerPort) RaiseCall() error {
	return p.tx.IssueCommand(p.context, &Raise{Call: true}, time.Now())
}

// UpgradeFirmware initiates an OTA firmware upgrade.
func (p *ServerPort) UpgradeFirmware(ota OTA) error {
	return p.tx.IssueCommand(p.context, &ota, time.Now())
}

// --- Lifecycle ---

// Close closes the port.
func (p *ServerPort) Close() error {
	p.cancel()
	p.input.Close()
	p.mixer.Close()

	// Wait for background goroutines to finish
	p.wg.Wait()

	// Safely close channels under lock to prevent send-to-closed-channel panic
	p.mu.Lock()
	p.closed = true
	close(p.statsChanges)
	close(p.stateEvents)
	p.mu.Unlock()

	return p.tx.Close()
}

// RecvFrom receives data from the given UplinkRx until closed or error.
// This method blocks; use `go port.RecvFrom(rx)` for non-blocking operation.
// Returns the first error encountered, or nil if all iterators completed normally.
func (p *ServerPort) RecvFrom(rx UplinkRx) error {
	defer rx.Close()

	var wg sync.WaitGroup
	var mu sync.Mutex
	var firstErr error

	setErr := func(err error) {
		if err == nil {
			return
		}
		mu.Lock()
		if firstErr == nil {
			firstErr = err
		}
		mu.Unlock()
	}

	wg.Add(3)
	go func() {
		defer wg.Done()
		for frame, err := range rx.OpusFrames() {
			if err != nil {
				setErr(err)
				return
			}
			p.HandleOpusFrames(frame)
		}
	}()
	go func() {
		defer wg.Done()
		for state, err := range rx.States() {
			if err != nil {
				setErr(err)
				return
			}
			p.HandleGearStateEvent(state)
		}
	}()
	go func() {
		defer wg.Done()
		for stats, err := range rx.Stats() {
			if err != nil {
				setErr(err)
				return
			}
			p.HandleGearStatsEvent(stats)
		}
	}()
	wg.Wait()
	return firstErr
}

const bufferingDuration = 360 * time.Millisecond

func (p *ServerPort) streamingOutputLoop() {
	str, err := opusrt.EncodePCMStream(
		p.mixer,
		p.mixer.Output().SampleRate(),
		p.mixer.Output().Channels(),
		p.tx.OpusEncodeOptions()...,
	)
	if err != nil {
		// Don't log error if context is already cancelled (expected shutdown)
		if p.context.Err() == nil {
			p.logger.ErrorPrintf("encode pcm stream: %v", err)
		}
		return
	}

	var stamp time.Time
	for {
		select {
		case <-p.context.Done():
			return
		default:
		}

		frame, duration, err := str.Frame()
		if err != nil {
			// Don't log error if context is cancelled (expected shutdown)
			if p.context.Err() == nil {
				p.logger.ErrorPrintf("read frame from mixer: %v", err)
			}
			return
		}

		now := time.Now()
		delay := stamp.Sub(now)

		if delay < 0 {
			// Behind schedule: reset timestamp
			stamp = now
		} else if delay < bufferingDuration {
			// Within buffering duration: fast buffering, minimal sleep
			time.Sleep(5 * time.Millisecond)
		} else {
			// Beyond buffering duration: sleep to maintain bufferingDuration buffer
			sleepDuration := delay - bufferingDuration
			time.Sleep(sleepDuration)
		}

		if len(frame) == 0 {
			stamp = stamp.Add(duration)
			continue
		}

		if err := p.tx.SendOpusFrames(p.context, opusrt.FromTime(stamp), frame.Clone()); err != nil {
			// Don't log error if context is cancelled (expected shutdown)
			if p.context.Err() == nil {
				p.logger.ErrorPrintf("send opus frame: %v", err)
			}
			continue
		}
		stamp = stamp.Add(frame.Duration())
	}
}

File: rust/chatgear/src/server_port.rs
//! Server port implementation.
//!
//! ServerPort manages server-side audio input/output, state/stats tracking, and device commands.

use crate::{
    ConnectedCellular, ConnectedWifi, DownlinkTx, GearStateEvent, GearStatsChanges,
    GearStatsEvent, PortError, ReadNFCTag, ServerPortRx, ServerPortTx, SessionCommand,
    SessionCommandEvent, StoredWifiList, UplinkRx, OTA,
    DeleteWifi, Halt, Raise, Reset, SetBrightness, SetLightMode, SetVolume, SetWifi,
};
use serde::Serialize;
use crate::port::AudioTrack;
use crate::port::AudioTrackCtrl;
use async_trait::async_trait;
use std::sync::Arc;
use tokio::sync::{mpsc, Mutex, RwLock};
use tokio_util::sync::CancellationToken;
use tracing::warn;

/// Server port for managing audio input/output and device state.
pub struct ServerPort<T: DownlinkTx> {
    tx: Arc<T>,
    cancel: CancellationToken,

    // Input - audio from device
    opus_frames_rx: Arc<Mutex<mpsc::Receiver<Vec<u8>>>>,
    opus_frames_tx: mpsc::Sender<Vec<u8>>,

    // Stats & State
    gear_stats: Arc<RwLock<Option<GearStatsEvent>>>,
    gear_state: Arc<RwLock<Option<GearStateEvent>>>,

    // Events
    stats_changes_rx: Arc<Mutex<mpsc::Receiver<GearStatsChanges>>>,
    stats_changes_tx: mpsc::Sender<GearStatsChanges>,
    state_events_rx: Arc<Mutex<mpsc::Receiver<GearStateEvent>>>,
    state_events_tx: mpsc::Sender<GearStateEvent>,

    closed: Arc<RwLock<bool>>,
}

impl<T: DownlinkTx + 'static> ServerPort<T> {
    /// Creates a new ServerPort.
    pub fn new(tx: T) -> Self {
        let (opus_tx, opus_rx) = mpsc::channel(1024);
        let (stats_tx, stats_rx) = mpsc::channel(32);
        let (state_tx, state_rx) = mpsc::channel(32);

        Self {
            tx: Arc::new(tx),
            cancel: CancellationToken::new(),
            opus_frames_rx: Arc::new(Mutex::new(opus_rx)),
            opus_frames_tx: opus_tx,
            gear_stats: Arc::new(RwLock::new(None)),
            gear_state: Arc::new(RwLock::new(None)),
            stats_changes_rx: Arc::new(Mutex::new(stats_rx)),
            stats_changes_tx: stats_tx,
            state_events_rx: Arc::new(Mutex::new(state_rx)),
            state_events_tx: state_tx,
            closed: Arc::new(RwLock::new(false)),
        }
    }

    /// Returns the cancellation token.
    pub fn cancellation_token(&self) -> CancellationToken {
        self.cancel.clone()
    }

    /// Returns the opus frames receiver.
    pub fn opus_frames_receiver(&self) -> Arc<Mutex<mpsc::Receiver<Vec<u8>>>> {
        self.opus_frames_rx.clone()
    }

    /// Returns the stats changes receiver.
    pub fn stats_changes_receiver(&self) -> Arc<Mutex<mpsc::Receiver<GearStatsChanges>>> {
        self.stats_changes_rx.clone()
    }

    /// Returns the state events receiver.
    pub fn state_events_receiver(&self) -> Arc<Mutex<mpsc::Receiver<GearStateEvent>>> {
        self.state_events_rx.clone()
    }

    /// Handles incoming opus frames from the device.
    pub async fn handle_opus_frames(&self, stamped_frame: Vec<u8>) {
        if let Err(e) = self.opus_frames_tx.send(stamped_frame).await {
            warn!("failed to buffer opus frame: {}", e);
        }
    }

    /// Handles incoming state events from the device.
    pub async fn handle_state_event(&self, event: GearStateEvent) {
        let closed = self.closed.read().await;
        if *closed {
            return;
        }

        // Filter out-of-order events
        {
            let current = self.gear_state.read().await;
            if let Some(ref current_state) = *current {
                if event.time < current_state.time {
                    return;
                }
            }
        }

        // Update state
        *self.gear_state.write().await = Some(event.clone());

        // Send event
        if let Err(e) = self.state_events_tx.send(event).await {
            warn!("state events channel full: {}", e);
        }
    }

    /// Handles incoming stats events from the device.
    pub async fn handle_stats_event(&self, event: GearStatsEvent) {
        let closed = self.closed.read().await;
        if *closed {
            return;
        }

        let changes = {
            let mut current = self.gear_stats.write().await;
            if let Some(ref mut current_stats) = *current {
                current_stats.merge_with(&event)
            } else {
                *current = Some(event);
                None
            }
        };

        if let Some(changes) = changes {
            if let Err(e) = self.stats_changes_tx.send(changes).await {
                warn!("stats changes channel full: {}", e);
            }
        }
    }

    /// Receives data from the given UplinkRx until closed.
    pub async fn recv_from<R: UplinkRx + 'static>(&self, rx: Arc<R>) {
        let cancel = self.cancel.clone();

        // Spawn opus frames receiver
        let rx_opus = rx.clone();
        let cancel_opus = cancel.clone();
        let opus_tx = self.opus_frames_tx.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = cancel_opus.cancelled() => break,
                    result = rx_opus.recv_opus_frame() => {
                        match result {
                            Ok(Some(frame)) => {
                                if opus_tx.send(frame).await.is_err() {
                                    break;
                                }
                            }
                            Ok(None) => break,
                            Err(_) => break,
                        }
                    }
                }
            }
        });

        // Spawn state receiver
        let rx_state = rx.clone();
        let cancel_state = cancel.clone();
        let gear_state = self.gear_state.clone();
        let state_tx = self.state_events_tx.clone();
        let closed = self.closed.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = cancel_state.cancelled() => break,
                    result = rx_state.recv_state() => {
                        match result {
                            Ok(Some(event)) => {
                                if *closed.read().await {
                                    break;
                                }
                                // Filter out-of-order
                                {
                                    let current = gear_state.read().await;
                                    if let Some(ref c) = *current {
                                        if event.time < c.time {
                                            continue;
                                        }
                                    }
                                }
                                *gear_state.write().await = Some(event.clone());
                                if state_tx.send(event).await.is_err() {
                                    break;
                                }
                            }
                            Ok(None) => break,
                            Err(_) => break,
                        }
                    }
                }
            }
        });

        // Spawn stats receiver
        let rx_stats = rx.clone();
        let cancel_stats = cancel.clone();
        let gear_stats = self.gear_stats.clone();
        let stats_tx = self.stats_changes_tx.clone();
        let closed = self.closed.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = cancel_stats.cancelled() => break,
                    result = rx_stats.recv_stats() => {
                        match result {
                            Ok(Some(event)) => {
                                if *closed.read().await {
                                    break;
                                }
                                let changes = {
                                    let mut current = gear_stats.write().await;
                                    if let Some(ref mut c) = *current {
                                        c.merge_with(&event)
                                    } else {
                                        *current = Some(event);
                                        None
                                    }
                                };
                                if let Some(changes) = changes {
                                    if stats_tx.send(changes).await.is_err() {
                                        break;
                                    }
                                }
                            }
                            Ok(None) => break,
                            Err(_) => break,
                        }
                    }
                }
            }
        });
    }

    /// Sends a command to the device.
    async fn send_command<C: SessionCommand + Serialize>(&self, cmd: &C) -> Result<(), PortError> {
        let event = SessionCommandEvent::new(cmd);
        self.tx
            .send_command(&event)
            .await
            .map_err(|e| PortError::CommandFailed(e.to_string()))
    }

    /// Closes the port.
    pub async fn close(&self) -> Result<(), PortError> {
        self.cancel.cancel();
        *self.closed.write().await = true;
        self.tx.close().await.map_err(|e| PortError::Other(e.to_string()))
    }
}

#[async_trait]
impl<T: DownlinkTx + 'static> ServerPortTx for ServerPort<T> {
    fn new_background_track(&self) -> Result<(Box<dyn AudioTrack>, Box<dyn AudioTrackCtrl>), PortError> {
        // Simplified - in real implementation, this would create mixer tracks
        Err(PortError::TrackCreationFailed("not implemented".to_string()))
    }

    fn new_foreground_track(&self) -> Result<(Box<dyn AudioTrack>, Box<dyn AudioTrackCtrl>), PortError> {
        Err(PortError::TrackCreationFailed("not implemented".to_string()))
    }

    fn new_overlay_track(&self) -> Result<(Box<dyn AudioTrack>, Box<dyn AudioTrackCtrl>), PortError> {
        Err(PortError::TrackCreationFailed("not implemented".to_string()))
    }

    fn background_track_ctrl(&self) -> Option<&dyn AudioTrackCtrl> {
        None
    }

    fn foreground_track_ctrl(&self) -> Option<&dyn AudioTrackCtrl> {
        None
    }

    fn overlay_track_ctrl(&self) -> Option<&dyn AudioTrackCtrl> {
        None
    }

    fn interrupt(&self) {
        // No-op in simplified implementation
    }

    async fn set_volume(&self, volume: i32) -> Result<(), PortError> {
        let cmd = SetVolume::new(volume);
        self.send_command(&cmd).await
    }

    async fn set_light_mode(&self, mode: &str) -> Result<(), PortError> {
        let cmd = SetLightMode::new(mode);
        self.send_command(&cmd).await
    }

    async fn set_brightness(&self, brightness: i32) -> Result<(), PortError> {
        let cmd = SetBrightness::new(brightness);
        self.send_command(&cmd).await
    }

    async fn set_wifi(&self, ssid: &str, password: &str) -> Result<(), PortError> {
        let cmd = SetWifi::new(ssid, "WPA2", password);
        self.send_command(&cmd).await
    }

    async fn delete_wifi(&self, ssid: &str) -> Result<(), PortError> {
        let cmd = DeleteWifi::new(ssid);
        self.send_command(&cmd).await
    }

    async fn reset(&self) -> Result<(), PortError> {
        let cmd = Reset::new();
        self.send_command(&cmd).await
    }

    async fn unpair(&self) -> Result<(), PortError> {
        let cmd = Reset::with_unpair();
        self.send_command(&cmd).await
    }

    async fn sleep(&self) -> Result<(), PortError> {
        let cmd = Halt::sleep();
        self.send_command(&cmd).await
    }

    async fn shutdown(&self) -> Result<(), PortError> {
        let cmd = Halt::shutdown();
        self.send_command(&cmd).await
    }

    async fn raise_call(&self) -> Result<(), PortError> {
        let cmd = Raise::call();
        self.send_command(&cmd).await
    }

    async fn upgrade_firmware(&self, ota: OTA) -> Result<(), PortError> {
        self.send_command(&ota).await
    }
}

impl<T: DownlinkTx + 'static> ServerPortRx for ServerPort<T> {
    fn opus_frames(&self) -> &mpsc::Receiver<Vec<u8>> {
        unimplemented!("use opus_frames_receiver() instead")
    }

    fn state_events(&self) -> &mpsc::Receiver<GearStateEvent> {
        unimplemented!("use state_events_receiver() instead")
    }

    fn stats_changes(&self) -> &mpsc::Receiver<GearStatsChanges> {
        unimplemented!("use stats_changes_receiver() instead")
    }

    fn gear_state(&self) -> Option<GearStateEvent> {
        // Blocking in async context - use gear_state_async in async code
        None
    }

    fn gear_stats(&self) -> Option<GearStatsEvent> {
        None
    }

    fn volume(&self) -> Option<i32> {
        None
    }

    fn light_mode(&self) -> Option<String> {
        None
    }

    fn brightness(&self) -> Option<i32> {
        None
    }

    fn wifi_network(&self) -> Option<ConnectedWifi> {
        None
    }

    fn wifi_store(&self) -> Option<StoredWifiList> {
        None
    }

    fn battery(&self) -> Option<(i32, bool)> {
        None
    }

    fn system_version(&self) -> Option<String> {
        None
    }

    fn cellular(&self) -> Option<ConnectedCellular> {
        None
    }

    fn pair_status(&self) -> Option<String> {
        None
    }

    fn read_nfc_tag(&self) -> Option<ReadNFCTag> {
        None
    }

    fn shaking(&self) -> Option<f64> {
        None
    }
}

// Async getters for ServerPort
impl<T: DownlinkTx + 'static> ServerPort<T> {
    /// Returns the current gear state (async version).
    pub async fn gear_state_async(&self) -> Option<GearStateEvent> {
        self.gear_state.read().await.clone()
    }

    /// Returns the current gear stats (async version).
    pub async fn gear_stats_async(&self) -> Option<GearStatsEvent> {
        self.gear_stats.read().await.clone()
    }

    /// Returns the current volume percentage.
    pub async fn volume_async(&self) -> Option<i32> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.volume.as_ref().map(|v| v.percentage as i32)
    }

    /// Returns the current light mode.
    pub async fn light_mode_async(&self) -> Option<String> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.light_mode.as_ref().map(|l| l.mode.clone())
    }

    /// Returns the current brightness percentage.
    pub async fn brightness_async(&self) -> Option<i32> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.brightness.as_ref().map(|b| b.percentage as i32)
    }

    /// Returns the current battery status (percentage, is_charging).
    pub async fn battery_async(&self) -> Option<(i32, bool)> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.battery.as_ref().map(|b| (b.percentage as i32, b.is_charging))
    }

    /// Returns the current WiFi network.
    pub async fn wifi_network_async(&self) -> Option<ConnectedWifi> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.wifi_network.clone()
    }

    /// Returns the shaking level.
    pub async fn shaking_async(&self) -> Option<f64> {
        let stats = self.gear_stats.read().await;
        stats.as_ref()?.shaking.as_ref().map(|s| s.level)
    }
}

#[cfg(test)]
mod server_port_tests {
    use super::*;
    use crate::conn_pipe::new_pipe;
    use crate::{GearState, UplinkTx, DownlinkRx};

    #[tokio::test]
    async fn test_server_port_send_command() {
        let (server, client) = new_pipe();
        let port = ServerPort::new(server);

        // Send set_volume command
        port.set_volume(75).await.unwrap();

        // Client receives it
        let received = client.recv_command().await.unwrap().unwrap();
        assert_eq!(received.cmd_type, "set_volume");
    }

    #[tokio::test]
    async fn test_server_port_receive_state() {
        let (server, client) = new_pipe();
        let port = ServerPort::new(server);

        // Client sends state
        let state = GearStateEvent::new(GearState::Recording);
        client.send_state(&state).await.unwrap();

        // Use handle_state_event directly
        port.handle_state_event(state.clone()).await;

        // Check state
        let current = port.gear_state_async().await.unwrap();
        assert_eq!(current.state, GearState::Recording);
    }

    #[tokio::test]
    async fn test_server_port_receive_stats() {
        let (server, _client) = new_pipe();
        let port = ServerPort::new(server);

        // Send stats
        let mut stats = GearStatsEvent::new();
        stats.volume = Some(crate::Volume {
            percentage: 80.0,
            update_at: giztoy_jsontime::Milli::now(),
        });

        port.handle_stats_event(stats).await;

        // Check volume
        let vol = port.volume_async().await.unwrap();
        assert_eq!(vol, 80);
    }

    #[tokio::test]
    async fn test_server_port_stats_merge() {
        let (server, _client) = new_pipe();
        let port = ServerPort::new(server);

        // First stats
        let mut stats1 = GearStatsEvent::new();
        stats1.volume = Some(crate::Volume {
            percentage: 50.0,
            update_at: giztoy_jsontime::Milli::now(),
        });
        port.handle_stats_event(stats1).await;

        // Second stats with battery
        let mut stats2 = GearStatsEvent::new();
        stats2.battery = Some(crate::Battery {
            percentage: 90.0,
            is_charging: true,
            ..Default::default()
        });
        port.handle_stats_event(stats2).await;

        // Both should be present
        let vol = port.volume_async().await.unwrap();
        assert_eq!(vol, 50);
        let (bat, charging) = port.battery_async().await.unwrap();
        assert_eq!(bat, 90);
        assert!(charging);
    }

    #[tokio::test]
    async fn test_server_port_all_commands() {
        let (server, client) = new_pipe();
        let port = ServerPort::new(server);

        // Test all command methods
        port.set_volume(50).await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "set_volume");

        port.set_brightness(80).await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "set_brightness");

        port.set_light_mode("dark").await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "set_light_mode");

        port.set_wifi("TestSSID", "password").await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "set_wifi");

        port.delete_wifi("TestSSID").await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "delete_wifi");

        port.reset().await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "reset");

        port.sleep().await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "halt");

        port.shutdown().await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "halt");

        port.raise_call().await.unwrap();
        let cmd = client.recv_command().await.unwrap().unwrap();
        assert_eq!(cmd.cmd_type, "raise");
    }

    #[tokio::test]
    async fn test_server_port_close() {
        let (server, _client) = new_pipe();
        let port = ServerPort::new(server);
        port.close().await.unwrap();
    }
}
